File: longestAbsoluteFilePath/longestAbsoluteFilePath.cpp
#include <iostream>
#include <stack>
#include <sstream>
using namespace std;

class Solution {
public:
    int lengthLongestPath(string input) {
        int res = 0;
        if (input.length() == 0)
            return res;
        istringstream iss(input);
        string line;
        stack<int> s;
        s.push(0);
        while (getline(iss, line, '\n')) {
            int level = line.find_last_of('\t') + 1;
            int len = line.length() - level;
            while (level + 1 < s.size()) {
                s.pop();
            }
            if (line.find('.') != string::npos) {
                res = max (res, s.top() + len);
            } else {
                s.push(s.top() + len + 1);
            }
        }
        return res;
    }
};

int main() {
  Solution s;
  cout << s.lengthLongestPath("dir\n\tsubdir1\n\t\tfile1.ext\n\t\tsubsubdir1\n\tsubdir2\n\t\tsubsubdir2\n\t\t\tfile2.ext") << endl;
  return 0;
}

File: longestAbsoluteFilePath/longestAbsoluteFilePath.java
class Solution {
    public int lengthLongestPath(String input) {
        Stack<Integer> stk = new Stack<>();
        int res = 0;
        for (String s : input.split("\n")) {
            int level = s.lastIndexOf("\t") + 1; // number of '\t', level example: 0,1,2,3...
            int len = s.length() - level + 1; //  +1 for '/'
            if (stk.empty()) { // level must be 0 here
                stk.push(len);
            } else {
                if (level == stk.size()) { // cur node in next level
                    int prevLen = stk.empty() ? 0 : stk.peek();
                    stk.push(len + prevLen);
                } else if (level == stk.size() - 1) { // cur node pre node in same level 
                    stk.pop();
                    int prevLen = stk.empty() ? 0 : stk.peek();
                    stk.push(len + prevLen);
                } else if (level < stk.size() - 1) {  // cur node in prev level
                    while (level < stk.size() - 1) stk.pop();
                    stk.pop();
                    int prevLen = stk.empty() ? 0 : stk.peek();
                    stk.push(len + prevLen);
                } else {
                    return -1;
                }
            }
            if (s.contains(".")) res = Math.max(res, stk.peek() - 1);
        }
        return res;
    }
}
